adam/perception/perception_frame_difference.py

```diff
@@ -0,0 +1,73 @@
+from abc import ABC
+from typing import Generic, TypeVar
+
+from attr import attrib, attrs
+from attr.validators import in_, instance_of
+from immutablecollections import ImmutableSet
+from adam.perception.developmental_primitive_perception import DevelopmentalPrimitivePerceptionFrame, ObjectPerception, \
+    PropertyPerception, RelationPerception
+
+
+class PerceptionFrameDiff(ABC):
+    r"""
+    Represents the difference between a `LanguageLearner`\ 's perception
+    of two different moments.
+    """
+
+
+PerceptionDiffT = TypeVar("PerceptionDiffT", bound="PerceptionFrameDiff")
+
+
+@attrs(slots=True, frozen=True)
+class DevelopmentalPrimitivePerceptionFrameDiff(PerceptionFrameDiff):
```

Comment:
It's often useful to make the distinction between "classes which compute things" and "classes which represent data".  I'd suggest restructuring this so that `DevelopmentalPrimitivePerceptionFrameDiff` holds not the before frame and after frame, but rather six fields `add_objects`, `removed_objects`, `added_property_assertions`, `removed_property_assertions`, `added_relations`, and `removed_relations`.  Then have a method `def diff_primitive_perception_frames(before: DevelopmentalPrimitivePerceptionFrame, after: DevelopmentalPrimitivePerceptionFrame) -> DevelopmentalPrimitivePerceptionFrameDiff`. 

Contrary to my earlier guidance, I think we can hold off on generics for now.